lib/datadog/appsec/contrib/devise/patcher.rb

```diff
@@ -23,16 +24,26 @@ def target_version
           end
 
           def patch
-            patch_authenticable_strategy
+            patch_authenticatable_strategy
+            patch_rememberable_strategy
             patch_registration_controller
 
             Patcher.instance_variable_set(:@patched, true)
           end
 
-          def patch_authenticable_strategy
+          def patch_authenticatable_strategy
+            ::Devise::Strategies::Authenticatable.alias_method(:__validate_datadog_authenticatable, :validate)
```

Comment:
This looks dangerous, as mixing aliasing and prepending can easily lead to obscure issues such as recursive loops.

AIUI the reason this was introduced is that the call stack is:

```
Foo#validate    # some app class that included ::Devise::Strategies::Rememberable and may implement validate that calls `super` or simply rely on it being provided by a parent
::Devise::Strategies::Rememberable#validate                                   # hypothetical, not there upstream as of yet
::Datadog::AppSec::Contrib::Devise::Patcher::AuthenticatablePatch#validate
::Devise::Strategies::Authenticatable#validate
```

But just prepending a patch results in this:

```
Foo#validate    # some app class that included ::Devise::Strategies::Rememberable and may implement validate that calls `super` or simply rely on it being provided by a parent
::Datadog::AppSec::Contrib::Devise::Patcher::RememberablePatch#validate
::Devise::Strategies::Rememberable#validate                                   # hypothetical, not there upstream as of yet
::Datadog::AppSec::Contrib::Devise::Patcher::AuthenticatablePatch#validate
::Devise::Strategies::Authenticatable#validate
```

But we want to skip some/all of `::Datadog::AppSec::Contrib::Devise::Patcher::AuthenticatablePatch#validate`, so this attempts to walk around it with:

```
Foo#validate    # some app class that included ::Devise::Strategies::Rememberable and may implement validate that calls `super` or simply rely on it being provided by a parent
::Datadog::AppSec::Contrib::Devise::Patcher::RememberablePatch#validate
::Devise::Strategies::Authenticatable#__validate_datadog_authenticatable
```

The problems are:

- suddenly `::Devise::Strategies::Authenticatable#validate` is not called at all (the method body is, but not the name). This can cause issues: a third party instrumenting `::Devise::Strategies::Authenticatable#validate` as well would never be called, or if `::Devise::Strategies::Authenticatable#validate`'s source code calls `super` it will actually look for `__validate_datadog_authenticatable` in a parent.
- Between `::Datadog::AppSec::Contrib::Devise::Patcher::RememberablePatch#validate` and `::Devise::Strategies::Authenticatable#__validate_datadog_authenticatable` the ancestor chain of calls is broken: Ruby "reenters" the instance and starts a new ancestor chain of calls, which creates the opportunity for infinite recursive looping.

I would favour a method where the patches collaborate through context passing: `RememberablePatch` sets a flag in a context, and `AuthenticatablePatch` checks for that flag.

Reply:
Thank you! It is indeed better to use some variable for context. Fixed in [c65e4b0](https://github.com/DataDog/dd-trace-rb/pull/3867/commits/c65e4b03ea961d2a1fd126e700e30e1999efeaad)